File: test-app/app/src/main/assets/app/tests/inheritanceChainResolutionTest.js
describe("Tests inheritance chain resolution", function () {
	
	var myCustomEquality = function(first, second) {
		return first == second;
	};
	
	beforeEach(function() {
		jasmine.addCustomEqualityTester(myCustomEquality);
	});
	
	it("TestCallWithStringArgumentWhenThereIsStringOverload", function () {
		
		__log("TEST: TestCallWithStringArgumentWhenThereIsStringOverload");
		
		var i = new com.tns.tests.InheritanceChainResolutionTest();
		
		var s = i.echo1("test123");
		
		expect(s).toBe("String=test123");
	});
	
	it("TestCallWithStringArgumentWhenThereIsNoStringOverload", function () {
		
		__log("TEST: TestCallWithStringArgumentWhenThereIsNoStringOverload");
		
		var i = new com.tns.tests.InheritanceChainResolutionTest();
		
		var s = i.echo2("test123");
		
		expect(s).toBe("Object=test123");
	});
});

File: test-app/app/src/main/java/com/tns/tests/InheritanceChainResolutionTest.java
package com.tns.tests;

import java.io.Serializable;

public class InheritanceChainResolutionTest {
    public InheritanceChainResolutionTest() {
    }

    public String echo1(Object s) {
        return "Object=" + s;
    }

    public String echo1(Serializable s) {
        return "Serializable=" + s;
    }

    public String echo1(CharSequence s) {
        return "CharSequence=" + s;
    }

    public String echo1(String s) {
        return "String=" + s;
    }

    public String echo2(Object s) {
        return "Object=" + s;
    }

    public String echo2(Serializable s) {
        return "Serializable=" + s;
    }

    public String echo2(CharSequence s) {
        return "CharSequence=" + s;
    }
}
